src/merkle_tree/constraints.rs

```diff
@@ -101,75 +45,253 @@ where
         let ns = cs.into();
         let cs = ns.cs();
         f().and_then(|val| {
-            let mut path = Vec::new();
-            for &(ref l, ref r) in val.borrow().path.iter() {
-                let l_hash = HGadget::OutputVar::new_variable(
-                    ark_relations::ns!(cs, "l_child"),
-                    || Ok(l),
+            let leaf_sibling = LeafH::OutputVar::new_variable(
+                ark_relations::ns!(cs, "leaf_sibling"),
+                || Ok(val.borrow().leaf_sibling_hash.clone()),
+                mode,
+            )?;
+            let leaf_position_bit = Boolean::new_variable(
+                ark_relations::ns!(cs, "leaf_position_bit"),
+                || Ok(val.borrow().leaf_index & 1 == 1),
+                mode,
+            )?;
+            let pos_list = val.borrow().position_list();
+            let mut path = Vec::with_capacity(pos_list.len() - 1);
+            for &bit in &pos_list[..pos_list.len() - 1] {
+                // we do not include leaf bit here, so ignore last element returned by pos_list
+                path.push(Boolean::new_variable(
+                    ark_relations::ns!(cs, "path_bit"),
+                    || Ok(bit),
                     mode,
```

Comment:
When the mode is `public` do we actually need to allocate the intermediate hashes as public?

I feel like the answer is no

/cc @Pratyush @weikengchen

Reply:
Is there a use case for requiring that the branch is a particular one? I guess not, right? If so, yeah, I think it makes sense to just allocate the root and the leaf hash/leaf

Reply:
The code block you referenced is actually the position of leaf. We do not store intermediate hashes in the path variable anyway (we only store the hashes in `auth_path`). Do we want to keep the position of `leaf` private in `public` mode? (Doing so, verifier might not be able to check if a leaf is at a particular position, right?) @ValarDragon